Fix typed parameters in IndexRequestBuilder and CreateIndexRequestBuilder

cbuescher opened the pull request:
IndexRequestBuilder#setSource as well as CreateIndexRequestBuilder#setSettings and CreateIndexRequestBuilder#setSouce will not work with `Map<String, String>` argument although the API looks like it should. This PR fixes the problem introducing correct wildcard parameters and adds tests.

Closes #10825

cbuescher committed:
Fix typed parameters in IndexRequestBuilder and CreateIndexRequestBuilder

IndexRequestBuilder#setSource as well as CreateIndexRequestBuilder#setSettings and
CreateIndexRequestBuilder#setSouce() will not work with Map<String, String> argument
although the API looks like it should. This PR fixes the problem introducing correct
wildcard parameters and adds tests.

Closes #10825

javanna commented on src/test/java/org/elasticsearch/action/index/IndexRequestBuilderTest.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class IndexRequestBuilderTest extends ElasticsearchSingleNodeTest {
```
do we really need a node here? I don't think we are sending any request? Should it just extend `ElasticsearchTestCase`?

cbuescher replied:
I had problems creating empty IndexRequestBuilder without a client (`IndexRequestBuilder(ElasticsearchClient client, IndexAction action)`), if there's any dummy client implementation I can pass in there that would probably possible. Used the ElasticsearchSingleNodeTest because of its handy creation of the IndexRequestBuilder.

cbuescher replied:
I found a workaround using an inner `DummyClient implements ElasticsearchClient` class, but that bloats the test. It's really fast though. Would be an ideal use case for Mockito or Jmock, not sure if I already asked around if we can use that in core? I think I saw it beeing used in Shield?

javanna replied:
I think we have at least one similar test that does the same, we can maybe share the dummy client to minimize the code repetition. It's not that we cannot use mockito, we could, we would need to have a bigger discussion around it, some people are in favour of that and some others are totally against it. I personally don't think mockito would solve all our problems, we should strive to make elasticsearch more unit testable, easier to say that to do though :)

javanna commented on src/test/java/org/elasticsearch/action/admin/indices/create/CreateIndexRequestBuilderTest.java:
```diff
@@ -0,0 +1,93 @@
+import java.util.HashMap;
+import java.util.Map;
+
+public class CreateIndexRequestBuilderTest extends ElasticsearchSingleNodeTest {
```
do we really need a node here? I don't think we are sending any request? Should it just extend `ElasticsearchTestCase`?

javanna commented:
left one comment around testing, looks good besides that

cbuescher commented:
I changed the tests to use an existing NoOpClient to make it a proper unit test case. To facilitate that I moved the formerly private test client to its own class alongside the original test it came from. This way I can use it in my own test setup.

javanna commented:
looks great thanks a lot @cbuescher
